src/Build/BackEnd/Shared/BuildResult.cs

```diff
@@ -598,8 +634,82 @@ void ITranslatable.Translate(ITranslator translator)
+
+            // This is a work-around for the bug https://github.com/dotnet/msbuild/issues/10208
+            // We are adding a version field to this class to make the ResultsCache backwards compatible with at least 2 previous releases.
+            // The adding of a version field is done without a breaking change in 3 steps, each separated with at least 1 intermediate release.
+            // 1st step (done): Add a special key to the dictionary. The presence of this key indicates that the version is serialized next.
+            // When serializing, add a key to the dictionary and a version field. Delete the special key from the dictionary during the deserialization and read a version if it presents.
+            // 2nd step: Stop writing a special key to the dictionary. Always serialize and de-serialize the version field. Remove the special keys if they present in the dictionary.
+            // 3rd step: Stop removing the special keys from the dictionary.
+            if (Traits.Instance.EscapeHatches.DoNotVersionBuildResult)
+            {
+                // Escape hatch: serialize/deserialize without version field.
+                translator.TranslateDictionary(ref _savedEnvironmentVariables, StringComparer.OrdinalIgnoreCase);
+                _version = 0;
+            }
+            else
+            {
+                lock (_lock)
+                {
+                    if (translator.Mode == TranslationDirection.WriteToStream)
+                    {
+                        // Add the special key VersionKeyName indicating the presence of a version to the _savedEnvironmentVariables dictionary.
+                        // If the dictionary was null, add another special key SavedEnvironmentVariablesDictionaryWasNull to the dictionary.
+                        if (_savedEnvironmentVariables is null)
+                        {
+                            _savedEnvironmentVariables = new Dictionary<string, string>
+                            {
+                                { SavedEnvironmentVariablesDictionaryWasNull, String.Empty }
+                            };
+                        }
+
+                        _savedEnvironmentVariables.Add(VersionKeyName, String.Empty);
+                        translator.TranslateDictionary(ref _savedEnvironmentVariables, StringComparer.OrdinalIgnoreCase);
+                        translator.Translate(ref _version);
+
+                        // Remove the added keys from the dictionary.
+                        if (_savedEnvironmentVariables.ContainsKey(SavedEnvironmentVariablesDictionaryWasNull))
+                        {
+                            _savedEnvironmentVariables = null;
+                        }
+                        else
+                        {
+                            _savedEnvironmentVariables.Remove(VersionKeyName);
+                        }
+                    }
+                    else
```

Comment:
Could you please specify which mode is here? Is it only read? are there any other modes ?

Reply:
There is only read and write modes there. So, one code path is reading, another is writing.

Reply:
Thank you for clarifying! Is there a chance to explicitly specify this mode per diffeerent else if (value == read mode )?